src/compute/src/render/context.rs

```diff
@@ -798,6 +1061,107 @@ where
+                specialized_arrangement_key(
+                    key.clone(),
+                    thinning.clone(),
+                    key_types.clone(),
+                    Vec::default(),
+                ),
+            );
+            let name = &format!("[32-bit] {}", name);
+            let oks = oks
+                .mz_arrange::<RowSpine<Bits32, Row, _, _>>(name, enable_arrangement_size_logging);
+            (SpecializedArrangement::Bits32Row(key_types, oks), errs)
+        } else {
+            // Catch-all: Just use RowRow.
+            let (oks, errs) = oks.map_fallible(
+                "FormArrangementKey",
+                specialized_arrangement_key(
+                    key.clone(),
+                    thinning.clone(),
+                    Vec::default(),
+                    Vec::default(),
+                ),
+            );
+            let oks =
+                oks.mz_arrange::<RowSpine<Row, Row, _, _>>(name, enable_arrangement_size_logging);
+            (SpecializedArrangement::RowRow(oks), errs)
+        }
+    }
+}
+
+/// Derives the column types of the key and values of an arrangement based on its column
+/// permutation, value thinning, and column types describing the full row schema.
+fn derive_key_val_types(
+    key: &Vec<MirScalarExpr>,
+    thinning: &Vec<usize>,
+    types: &Vec<ColumnType>,
+) -> (Vec<ColumnType>, Vec<ColumnType>) {
+    let mut key_types = Vec::new();
+    let mut val_types = Vec::new();
+    if types.len() > 0 {
+        for i in 0..key.len() {
+            key_types.push(key[i].typ(types).clone());
+        }
+        for c in thinning.iter() {
+            val_types.push(types[*c].clone());
+        }
+    }
+    (key_types, val_types)
+}
+
+/// Obtains a function that maps input rows to (key, value) pairs according to
+/// the given key and thinning expressions. This function allows for specialization
+/// of key and value types and is intended to use to form arrangement keys.
+fn specialized_arrangement_key<K, V>(
+    key: Vec<MirScalarExpr>,
+    thinning: Vec<usize>,
+    key_types: Vec<ColumnType>,
+    val_types: Vec<ColumnType>,
+) -> impl FnMut(Row) -> Result<(K, V), DataflowError>
+where
+    K: Columnation + Data + FromRowByTypes,
+    V: Columnation + Data + FromRowByTypes,
+{
+    let mut row_buf = Row::default();
+    let mut datums = DatumVec::new();
+    move |row| {
+        // TODO: Consider reusing the `row` allocation; probably in *next* invocation.
+        let datums = datums.borrow_with(&row);
+        let temp_storage = RowArena::new();
+        row_buf
+            .packer()
+            .try_extend(key.iter().map(|k| k.eval(&datums, &temp_storage)))?;
+        let key_row = row_buf.clone();
+        row_buf.packer().extend(thinning.iter().map(|c| datums[*c]));
+        let val_row = row_buf.clone();
+        Ok::<(K, V), DataflowError>((
+            K::from_row(key_row, &key_types),
+            V::from_row(val_row, &val_types),
+        ))
+    }
```

Comment:
I think we can avoid the `clone` calls here by passing `&row_buf` to `from_row`

Reply:
An improved version based on a new `from_datum_iter` method is now in. It avoids the clones here, but of course we must still allocate the target type (`Row`, `Bytes9`) for the result.

Reply:
Relatedly, I also now included a final commit avoiding row encoding/decoding for `BytesN` keys in `CollectionBundle::flat_map`.